### PR TITLE
fix!: mn_rr features only for v21+

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
Should not be 2 forks in one version

## What was done?
 - Asset Unlock transactions (withdrawals) should be available only in MN_RR fork
 - MN_RR should not be auto-activated on Main net without intentional release of code (and not by spork), but they are need on test net to test platform.

## How Has This Been Tested?
Run unit/functional tests

## Breaking Changes
Yes (see "what was done")

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/evo/specialtxman.cpp [hunk 1]
```diff
@@ -50,11 +50,17 @@ static bool CheckSpecialTxInner(const CTransaction& tx, const CBlockIndex* pinde
...
             if (!llmq::utils::IsV20Active(pindexPrev)) {
                 return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetlocks-before-v20");
             }
             return CheckAssetLockUnlockTx(tx, pindexPrev, indexes, state);
+        case TRANSACTION_ASSET_UNLOCK:
+            if (!llmq::utils::IsMNRewardReallocationActive(pindexPrev)) {
+                if (Params().NetworkIDString() != CBaseChainParams::REGTEST || !llmq::utils::IsV20Active(pindexPrev)) {
+                    return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetunlocks-before-mn_rr");
+                }
+            }
```
Comment (PastaPastaPasta) [inline src/evo/specialtxman.cpp:62]: I feel like this logic is wrong; or at least it feels confusing; change to be
```
if (net_id == regtest && !v20Active) return false;
if (!mn_rr_active) return false;
return checkassetunlock()
```

File: src/llmq/ehf_signals.cpp [hunk 2]
```diff
@@ -91,6 +95,10 @@ void CEHFSignalsHandler::trySignEHFSignal(int bit, const CBlockIndex* const pind
 
 void CEHFSignalsHandler::HandleNewRecoveredSig(const CRecoveredSig& recoveredSig)
 {
+    if (Params().NetworkIDString() == CBaseChainParams::MAIN) {
+        // TODO: Only one fork on MainNet in v20. These changes are may be activated only in v21
```
Comment (PastaPastaPasta) [inline src/llmq/ehf_signals.cpp:99]: same

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge; this looks correct to me

**Comment by UdjinM6** [comment]
>MN_RR should not be auto-activated on Main net without intentional release of code **_(and not by spork)_**, but they are need on test net to test platform

so is `SPORK_24_EHF` not longer need or is it a non-mainnet only spork now?

**Comment by PastaPastaPasta** [comment]
> > MN_RR should not be auto-activated on Main net without intentional release of code **_(and not by spork)_**, but they are need on test net to test platform
> 
> so is `SPORK_24_EHF` not longer need or is it a non-mainnet only spork now?

Basically non-main net only, a value could still be set; but it doesn't (shouldn't) affect anything on mainnet

File: src/evo/specialtxman.cpp [hunk 3]
```diff
@@ -50,11 +50,18 @@ static bool CheckSpecialTxInner(const CTransaction& tx, const CBlockIndex* pinde
...
                 return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetlocks-before-v20");
             }
             return CheckAssetLockUnlockTx(tx, pindexPrev, indexes, state);
+        case TRANSACTION_ASSET_UNLOCK:
+            if (Params().NetworkIDString() == CBaseChainParams::REGTEST && !llmq::utils::IsV20Active(pindexPrev)) {
+                return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetunlocks-before-v20");
+            }
+            if (Params().NetworkIDString() != CBaseChainParams::REGTEST && !llmq::utils::IsMNRewardReallocationActive(pindexPrev)) {
+                return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetunlocks-before-mn_rr");
+            }
```
Comment (UdjinM6) [inline src/evo/specialtxman.cpp:63]: why can't we just do
```suggestion
            if (!llmq::utils::IsMNRewardReallocationActive(pindexPrev)) {
                return state.Invalid(TxValidationResult::TX_CONSENSUS, "assetunlocks-before-mn_rr");
            }
```
and adjust tests accordingly instead?

**Comment by knst** [comment]
> so is SPORK_24_EHF not longer need or is it a non-mainnet only spork now?

This spork has been introduced to prevent activation of mn_rr before platform is ready.

As discussed today:
 - Asset Unlock transaction is not tested properly on testnet yet; not tested properly feature should not be potentially active on mainnet; also it may require consensus rules to chage
 - 2 forks are confusing for dash users. Let's keep the rule "1 major version = 1 hard-fork"
 - activation of a hard fork by spork is it bad idea; because it should relay on user's update, not DCG command to do anything.

So, mn_rr requires a new release v21 or v20.1; the spork maybe even dropped then.

For testnet it is needed for platform's need + to test EHF mechanism.

File: src/evo/specialtxman.cpp (hunk 3 above)
Comment (knst) [reply src/evo/specialtxman.cpp:63]: > and adjust tests accordingly instead?

non-trivial change due to withdrawal limits.

Comment (PastaPastaPasta) [reply src/evo/specialtxman.cpp:63]: We probably should in future PRs; this is intended to just quickly resolve this.

**Review by UdjinM6 — APPROVED** [review]
utACK

File: src/evo/specialtxman.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/evo/specialtxman.cpp]: Completely agree; but here we're just changing which version activates it on regtest; not really actually changing any functionality. But I do completely agree this should be changed and fix tests before rc.2

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**